eZ/Publish/Core/Repository/Helper/RoleDomainMapper.php

```diff
@@ -199,6 +201,29 @@ public function buildPersistenceRoleCreateStruct(APIRoleCreateStruct $roleCreate
         );
     }
 
+    /**
+     * Creates SPI Role copy struct from provided API role copy struct.
+     */
+    public function buildPersistenceRoleCopyStruct(APIRoleCopyStruct $roleCopyStruct, int $clonedId): SPIRoleCopyStruct
+    {
+        $policiesToCopy = [];
+        foreach ($roleCopyStruct->getPolicies() as $policyCopyStruct) {
+            $policiesToCopy[] = $this->buildPersistencePolicyObject(
+                $policyCopyStruct->module,
+                $policyCopyStruct->function,
+                $policyCopyStruct->getLimitations()
+            );
+        }
```

Comment:
This piece of code actually made clear to me that `RoleCopyStruct` should extend `RoleCreateStruct`. It's exactly the same as in `buildPersistenceRoleCreateStruct`. Moreover comparing `RoleCopyStruct` and `RoleCreateStruct`, both contain almost the same properties. On a DDD/OOP level it also feels right (as copying is a more automated process of creating).
Then, you could extract this code to a common method accepting `RoleCreateStruct`.